<a href="https://colab.research.google.com/github/AdamStamm/bootcamp_alura/blob/main/Notebooks/Bootcamp_Data_Science_Modulo_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Apresentação

Esse notebook contempla os códigos aprendidos no **Módulo 3** do Bootcamp Data Science aplicada desenvolvido pela [Alura](http://alura.com.br). O foco de aprendizado é **Análise descritiva e testes estatíticos**.

#Importando bibliotecas.

In [1]:
import pandas as pd
from zipfile import ZipFile
 
#biblioteca de fluxo de entrada e saída
#importamos o pacote de leitura binária
from io import BytesIO
import requests

- A biblioteca IO foi feita para controle de fluxos de entrada/saída de dados, no módulo anterior usamos o pacote StringIO da io para capturar o fluxo de dados que estavam na memória (copy and paste). Agora usaremos para capturar o fluxo de dados binários.
Mais informações: https://docs.python.org/3/library/io.html
 
- A biblioteca requests permite que você faça requisições HTTP de maneira fácil sem a necessidade de uma API.
mais informações [aqui](https://requests.readthedocs.io/pt_BR/latest/user/quickstart.html) ou [aqui](https://www.digitalocean.com/community/tutorials/how-to-get-started-with-the-requests-library-in-python-pt)

#Aula 1 - Educação, saúde e tipos de dados.

##Carregando dados

Dessa vez carregaremos os dados de um arquivo zipado, sem descompacta-lo no computador, por maior comodidade upei o arquivo zip direto no meu [Github](https://github.com/AdamStamm/bootcamp_alura/).

In [2]:
#Fazendo requisição binária do arquivo zip do github
path = 'https://github.com/AdamStamm/bootcamp_alura/blob/main/dados/PeNSE_2015_AMOSTRA2.zip?raw=true'
r = requests.get(path)
zip_file = BytesIO(r.content)
 
# Leitura do arquivo zip
zip = ZipFile(zip_file) 
 
#pega o caminho da lista de arquivos
caminho_arquivo_zip = zip.namelist()[1]
 
df = pd.read_csv(zip.open(caminho_arquivo_zip), sep=';')

In [3]:
df.head()

ANOPESQ  PAIS  REGEOGR  VB00004  VB01001  ...  V0008  V0041  aluno  escola  turma
0     2015    76        1        1        2  ...      4      1      1       1    296
1     2015    76        1        1        2  ...      4      1      2       1    296
2     2015    76        1        1        1  ...      4      1      3       1    296
3     2015    76        1        1        1  ...      4      1      4       1    296
4     2015    76        1        1        1  ...      4      1      5       1    296

[5 rows x 181 columns]

Algumas propriedades do ZipFile

In [4]:
#mostrando o objeto que o ZipFile possui.  
zip

<zipfile.ZipFile file=<_io.BytesIO object at 0x7f331a816bf8> mode='r'>

In [5]:
#vendo a lista de arquivos contidos no .zip
zip.infolist()

[<ZipInfo filename='arquivos csv/' external_attr=0x10>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV' compress_type=deflate external_attr=0x20 file_size=7262170 compress_size=1235667>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=11420533 compress_size=1329882>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=5000642 compress_size=114324>,
 <ZipInfo filename='arquivos sas/' external_attr=0x10>,
 <ZipInfo filename='arquivos sas/pense_amostra2_aluno.sas7bdat' compress_type=deflate external_attr=0x20 file_size=24691712 compress_size=1957530>,
 <ZipInfo filename='arquivos sas/pense_amostra2_alunoescola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=45253632 compress_size=2461932>,
 <ZipInfo filename='arquivos sas/pense_amostra2_escola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=18105344 compress_size=332162>,
 <Zip

In [6]:
#outra forma de ver os arquivos, retornando uma lista
zip.namelist()

['arquivos csv/',
 'arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV',
 'arquivos sas/',
 'arquivos sas/pense_amostra2_aluno.sas7bdat',
 'arquivos sas/pense_amostra2_alunoescola.sas7bdat',
 'arquivos sas/pense_amostra2_escola.sas7bdat',
 'arquivos xlsx/',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNO.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNOESCOLA.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ESCOLA.xlsx',
 'Dicionario_PENSE_Microdados_AMOSTRA2.xls']

Propriedade do requests

In [7]:
#verificando a URL que a requests capturou
r.url

'https://raw.githubusercontent.com/AdamStamm/bootcamp_alura/main/dados/PeNSE_2015_AMOSTRA2.zip'

Iniciando análises iniciais

In [8]:
#verificando o tamanho do data frame
print(f'O conjunto tem {df.shape[0]} linhas e {df.shape[1]} colunas')

O conjunto tem 16556 linhas e 181 colunas


In [9]:
#visualizando colunas
df.columns

Index(['ANOPESQ', 'PAIS', 'REGEOGR', 'VB00004', 'VB01001', 'VB01002',
       'VB01003', 'VB01004', 'VB01005', 'VB01006',
       ...
       'ESTRATO_EXP', 'ESTRATOGEOREG', 'PESO', 'V0006', 'V0007', 'V0008',
       'V0041', 'aluno', 'escola', 'turma'],
      dtype='object', length=181)

Percebemos que existem colunas com os nomes das variáveis difíceis de serem identificados, por essa razão é necessário estudarmos o dicionário de variáveis que acompanha o arquivo zip. Uma outra forma de abordagem é observar a forma desses dados como faremos a seguir.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16556 entries, 0 to 16555
Columns: 181 entries, ANOPESQ to turma
dtypes: float64(7), int64(171), object(3)
memory usage: 22.9+ MB


Percebemos que existem 7 colunas do tipo float, 171 do tipo inteiro e 3 objects.

In [11]:
#visualizando dados por tipo
df.select_dtypes(include='object')

VB17003 VB17004        PESO
0        37,1     157  299,735235
1        50,2     152  355,170081
2        48,2   161,4  299,735235
3        69,4   172,5  355,170081
4        78,9   167,5  299,735235
...       ...     ...         ...
16551    56,7   167,9  447,815539
16552    61,9   185,8   592,50744
16553    60,9   159,4  401,131512
16554    47,3   157,8  442,718091
16555    62,6   163,9   592,50744

[16556 rows x 3 columns]

Note que o tipo object em geral trata de strings, mas temos 3 colunas de tipo float sendo tratadas como strings, isso acontece por que no nosso idioma pontos flutuantes são debitados por vírgula, no formato americano são demorados por ponto, isso faz com que o código seja classificado como string na hora do carregamento dos dados. Para corrigir isso basta, na hora do carregamento dos dados, informar essa característica incluindo os parâmetros thousands para milhar e decimal para decimais.
 
Também é importante olhar o dicionário de dados, se verificarmos a coluna PESO refere-se ao peso *estatístico* dos dados e **não** do peso dos alunos.

- **VB17003:** Peso(Kg) do aluno
- **VB17004:** Altura do aluno
- **PESO:** Peso amostral para expansão.

In [12]:
#carregando conjunto de dados com os parâmetros adicionais (deixei uma nova
#célula apenas por questão didática)
df = pd.read_csv(zip.open(caminho_arquivo_zip), sep=';',thousands='.', decimal=',')

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16556 entries, 0 to 16555
Columns: 181 entries, ANOPESQ to turma
dtypes: float64(10), int64(171)
memory usage: 22.9 MB


Note que já não temos mais as colunas do tipo objeto, vamos observar então as colunas do tipo float.

In [14]:
df.select_dtypes(include='float64')

TEMPODESLOC  TEMPOTOTAL  TEMPOEDFIS  ...  VB17003  VB17004        PESO
0              0.0        15.0         0.0  ...     37.1    157.0  299.735235
1            120.0       510.0       260.0  ...     50.2    152.0  355.170081
2              0.0       585.0       130.0  ...     48.2    161.4  299.735235
3              0.0         0.0         0.0  ...     69.4    172.5  355.170081
4             30.0       165.0        90.0  ...     78.9    167.5  299.735235
...            ...         ...         ...  ...      ...      ...         ...
16551         30.0       400.0       110.0  ...     56.7    167.9  447.815539
16552         20.0       800.0       325.0  ...     61.9    185.8  592.507440
16553          0.0         0.0         0.0  ...     60.9    159.4  401.131512
16554         75.0       215.0        35.0  ...     47.3    157.8  442.718091
16555         50.0       485.0       110.0  ...     62.6    163.9  592.507440

[16556 rows x 10 columns]

*Dicionário:

- **VB17001**: Qual o peso do aluno?
- **VB17002**: Qual a altura do aluno?

OBS: Anteriormente tinhamos visto que peso e aluna do aluno eram a coluna VB17003 e VB17004, por que isso? As de final 01 e 02 são dados aferidos dos alunos e as 03 e 04 são dados imputados pelo sistema.

#Tipos de variáveis.

##Exemplos de tipos distintos

**Qualitativas**: São características que não possuem valores quantitativos, mas, ao contrário são definidas por diversas categorias, ou seja, representam uma classificação dos indivíduos. Podem ser de dois tipos.

- **Nominais**: São dados que não podem ser ordenados entre as categorias. Ex: cor dos olhos, doente/sadio, etc...

- **Ordinais**: São dados que possuem uma ordenação entre as categorias. Ex: estágio da doença (inicial, intermediário, terminal), mês de observação (janeiro, fevereiro...).

**Quantitativas**: São as características que podem ser medidas em uma escala quantitativa, ou seja, apresentam valores numéricos que fazem sentido. Também podem ser de dois tipos.

- **Contínuas**:  Características mensuráveis que assumem valores em uma escala contínua (na reta real do plano cartesiano), para quais valores fracionais fazem sentido. Ex: Peso (balança), altura (régua), tempo (relógio), pressão arterial, idade, etc...

- **Discretas**: São características mensuráveis que podem assumir apenas um número finito ou infinito contável de valores e, assim, somente fazem sentido valores inteiros. Geralmente são o resultado de contagens. Ex: Número de filhos, número de bactérias por litro de leite, número de cigarros fumados por dia.

#Desafio 1: Descompactar a pasta zipada utilizando outro método.

Para esse desafio irei fazer upload da pasta para o Google drive uma vez que o GitHub não permite a manipulação de dados sem que sejam dados commits. Portanto irei montar o drive e então trabalhar com o zip.
 
Após o drive montado irei utilizar comandos de Shell Script, ou seja, comandos bashrc. Os servidores colab utilizam por baixo dos panos o sistema Unix e isso permite esse tipo de utilização. Para enviarmos comandos para o sistema usamos o sinal de exclamação antes do comando.

In [15]:
!unzip /content/drive/MyDrive/Dados_Bootcamp/PeNSE_2015_AMOSTRA2.zip -d /content/drive/MyDrive/Dados_Bootcamp

Archive:  /content/drive/MyDrive/Dados_Bootcamp/PeNSE_2015_AMOSTRA2.zip
   creating: /content/drive/MyDrive/Dados_Bootcamp/arquivos csv/
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV  
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV  
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV  
   creating: /content/drive/MyDrive/Dados_Bootcamp/arquivos sas/
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos sas/pense_amostra2_aluno.sas7bdat  
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos sas/pense_amostra2_alunoescola.sas7bdat  
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos sas/pense_amostra2_escola.sas7bdat  
   creating: /content/drive/MyDrive/Dados_Bootcamp/arquivos xlsx/
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos xlsx/PENSE_AMOSTRA2_ALUNO.xlsx  
  inflating: /content/drive/MyDrive/Dados_Bootcamp/arquivos xlsx

In [16]:
#visualizando arquivos extraídos
!ls /content/drive/MyDrive/Dados_Bootcamp

'arquivos csv'	  Dicionario_PENSE_Microdados_AMOSTRA2.xls
'arquivos sas'	  PeNSE_2015_AMOSTRA2.zip
'arquivos xlsx'


#Desafio 2: Estudar a base PeNSE.

In [17]:
#incluindo vídeo no notebook
from IPython.core.display import HTML
 
HTML('<iframe id="ytplayer" type="texto/html" width="500" height="300" \ src="https://www.youtube.com/embed/watch?v=z4BBCHZUvD4" frameborder="2"/>')

Os dados que serão trabalhados são da Pesquisa Nacional de Saúde do Escolar de 2015. Essa foi a terceira edição da PeNSE e foi a primeira que inclui alunos do ensino médio na faixa dos 13 aos 17 anos, as edições anteriores contemplavam apenas alunos até o nono ano (antiga oitava série).
 
Os estudantes são selecionados de um conjunto de escolas previamente definido (Amostra), e para cada um desses grupos escolares foi criado um plano tabular.
 
A Pesquisa é realizada 
graças à parceria do Ministério da Saúde com o Instituto Brasileiro de Geografia e Estatística e o apoio do Ministério da 
Educação. Ela faz parte da Vigilância dos Fatores de Risco e Proteção das Doenças Crônicas do Brasil. Em 2015, com a inclusão da amostra de estudantes de 13 a 17 anos, tornou a PeNSE 
comparável aos inquéritos internacionais.

#Desafio 3: Classificar algumas das perguntas do conjunto de dados como qualitativas e quantitativas.

**Quantitativas**
- Discretas:
 
>>- VB01010A (Contando com você, quantas pessoas moram na sua casa ou apartamento?)
- VB01019 (Quantos banheiros com chuveiro têm dentro da sua casa?)
- VB02001 (Nos últimos 7 dias, em quantos dias você comeu feijão?)
- VB02002 (Nos últimos 7 dias, em quantos dias você comeu salgados fritos?)
- VB02004A (Nos últimos 7 dias, em quantos dias você comeu pelo menos um tipo de vegetal?)
- VB03001A2 (Nos últimos 7 dias, em quantos você voltou a pé ou de bicicleta da escola?)

- Contínuas:

>>- VB17001: Qual o peso do aluno?
- VB17002: Qual a altura do aluno?
- TEMPODESLOC: Tempo de deslocamento diário acumulado pelo escolar, com deslocamento da casa para a escola e da escola para casa feito a pé ou de bicicleta.
- TEMPOEDFIS: Tempo médio acumulado em que o escolar fez atividades físicas.
- TEMPOEXTRA: Tempo médio diário acumulado do escolar com a prática de uma atividade física extraescolar como esportes, dança, ginástica, musculação...


**Qualitativas**
- Nominais:

>>- VB01023 (Você estuda em regime integral?)
- VB01024 (Você estuda em regime de internato?)
- VB01026 (Quando terminar o ciclo/curso que você está frequentando atualmente, você pretende?)
- VB01006 (Você mora com sua mãe?)
- VB01007 (Você mora com seu pai?)
- VB01013 (Na sua casa tem telefone fixo?)
- VB01014 (Você tem celular?)
- VB01015A (Na sua casa tem computador?)
- VB01016 (Você tem aceso à internet em sua casa?)
- VB01018 (Alguém que mora na sua casa tem carro?)

- Ordinais:

>>- VB01004 (Mês de aniversário?)
- VB01005 (Em que ano nasceu?)
- VB01021 (Em que série/ano você está?)
- VB01022 (Em que turno você estuda?)
- VB01025 (Qual o grau de escolaridade mais elevado que pretendo concluir?)
- VB01008A (Que nível de ensino (grau) sua mãe estudou ou estuda?
- VB03002A1 (Quando você vai para a escola a pé ou de bicicleta, quanto tempo você gasta?
- VB06006 (Quantos amigos seus usam drogas?)